index.js

```diff
@@ -1,9 +1,87 @@
-var crypto = require('crypto')
+var crypto = require('crypto');
+var fork = require('child_process').fork;
+var path = require('path');
+var compat = require('./browser');
+var hashes = {
+  sha1: 0,
+  0: 'sha1',
+  sha224: 1,
+  1: 'sha224',
+  sha256: 2,
+  2: 'sha256',
+  sha384: 3,
+  3: 'sha384',
+  sha512: 4,
+  4: 'sha512'
+};
+function asyncpbkdf2(password, salt, iterations, keylen, digest, callback) {
+  if (typeof iterations !== 'number') {
+    throw new TypeError('Iterations not a number')
+  }
+  if (iterations < 0) {
+    throw new TypeError('Bad iterations')
+  }
 
-var exportFn = require('./pbkdf2')
-var exported = exportFn(crypto)
+  if (typeof keylen !== 'number') {
+    throw new TypeError('Key length not a number')
+  }
 
-module.exports = {
-  pbkdf2: exported.pbkdf2,
-  pbkdf2Sync: exported.pbkdf2Sync
+  if (keylen < 0) {
+    throw new TypeError('Bad key length')
+  }
+  var msg = {
+    password: password.toString(),
+    salt: salt.toString(),
+    iterations: iterations,
+    keylen: keylen,
+    digest: digest
+  };
+
+  var child = fork(path.resolve(__dirname, 'async-shim.js'));
+  child.on('message', function (resp) {
+    child.kill();
+    callback(null, new Buffer(resp,'hex'));
+  }).on('error', function (err) {
+    child.kill();
+    callback(err);
+  });
+  child.send(msg);
+}
+exports.pbkdf2 = pbkdf2;
+exports.pbkdf2Sync = pbkdf2Sync;
+function pbkdf2Sync(password, salt, iterations, keylen, digest) {
+  digets = digest || 'sha1';
+  if (isNode10()) {
+    if (digest === 'sha1') {
+      return crypto.pbkdf2Sync(password, salt, iterations, keylen);
+    } else {
+      return compat.pbkdf2Sync(password, salt, iterations, keylen, digest);
+    }
+  } else {
+    return crypto.pbkdf2Sync(password, salt, iterations, keylen, digest);
+  }
+}
+function pbkdf2(password, salt, iterations, keylen, digest, callback) {
+  if (typeof digest ===  'function') {
+    callback = digest;
+    digest = 'sha1';
+  }
+
+  if (isNode10()) {
+    if (digest === 'sha1') {
+      return crypto.pbkdf2(password, salt, iterations, keylen, callback);
+    } else {
+      return asyncpbkdf2(password, salt, iterations, keylen, digest, callback);
+    }
+  } else {
+    return crypto.pbkdf2(password, salt, iterations, keylen, digest, callback);
+  }
 }
+var sha1 = '0c60c80f961f0e71f3a9b524af6012062fe037a6e0f0eb94fe8fc46bdc637164';
+var isNode10Result;
+function isNode10() {
```

Comment:
i mean yes it will slow down the very first time you create one but not by much.  Open to smarter ways to figure this out.